alternator/streams.cc

```diff
@@ -224,38 +226,59 @@ future<alternator::executor::request_return_type> alternator::executor::list_str
 
 struct shard_id {
     static constexpr auto marker = 'H';
-
-    utils::UUID table;
-    db_clock::time_point time;
-    cdc::stream_id id;
-
-    shard_id(utils::UUID tab, db_clock::time_point t, cdc::stream_id i)
-        : table(tab)
-        , time(t)
-        , id(i)
-    {}
+    static constexpr auto hex_width = sizeof(uint64_t)*8/4;
+
+    db_clock::time_point start_time;
+    db_clock::time_point end_time; // zero if open
+    dht::token lo_token; // inclusive
+    dht::token hi_token; // exclusive
```

Comment:
Yes. lo_token == start of range (inclusive), hi_token == end of range (exclusive). hi_token is typically the next shards lo_token.

Reply:
So why is end of range exclusive here, if vnodes have the end of range inclusive? Maybe it doesn't matter for correctness, but the inconsistency is weird.

Reply:
Unless the `hi_token` here is the vnode's end + 1, and `lo_token` is the vnode's start - 1, but then it's even more weird.

Reply:
Because it is based on info in stream_id:s. lo_token is the first (start) token of the range determined by the set of ID:s having index N, and hi_token is the first token of N+1. Thus the (min, max] range.

Reply:
However, if I misunderstood the code in generations.cc so that a stream_id:s token is an exclusive start, then the comment and the selects in get_records need adjusting, otherwise they will be subtly ineffective.

Reply:
There may be a problem.

The token inside stream_id is not supposed to describe neither the start nor the end of any vnode, it is just defined so that all IDs with the same index fall into the same vnode in such a way that the token of each ID is mapped to a particular shard on one of the nodes.

So if you define ranges using the tokens of stream IDs by taking token of one stream ID in each group as a boundary, they will cover the entire token ring, but they won't match vnodes. You'll get something like the following picture:
![image](https://user-images.githubusercontent.com/6516501/94041936-30f87880-fdcb-11ea-89fd-4ef4772fc5b4.png)
where the green color denotes vnode boundaries, and red color denotes your boundaries that you defined using stream IDs.

The information about vnode boundaries is contained inside the internal description table, where generations are lists of tuples of the following type:
```
struct token_range_description {
    dht::token token_range_end;
    std::vector<stream_id> streams;
    uint8_t sharding_ignore_msb;
```
the `token_range_end` is the info we're looking for. It is not guaranteed that any of the streams' IDs contains `token_range_end` as its token.

But not in the user-facing table which has just a set of streams, this information is lost.

Reply:
We are not really interested in the actual vnode range as such, we are interested in a range such that all the ID:s for this vnode are inside it. Using lowest == inclusive start, and next ID start as exclusive high should fulfill this criteria. 
It could be a problem though if the query will still consider querying _into_ another vnode (though no ID:s should match), i.e. the lowest "start" of ID:s for the "next" vnode is "higher" than vnode range start. 
But ID token is defined as "first token for (scylla) shard" of a given vnode. Logically it would seem that you cannot have a token lower than this that hits the same shard (and by extension vnode).

Reply:
> It could be a problem though if the query will still consider querying into another vnode

I'd assume that range scans don't know whether or not there will exist a partition with data in the other node, so it will have to look into it. Btw. why the decision to use partition scans and not IN queries?

> But ID token is defined as "first token for (scylla) shard" of a given vnode. Logically it would seem that you cannot have a token lower than this that hits the same shard (and by extension vnode).

The vnode may start inside shard 1 say, to the smallest stream ID, which is defined to be the "first token for shard 0", will not be on the vnode boundary.

Reply:
Your code in generation splits the ring into vnodes by token boundaries, then for each such boundary range find all shard boundaries and generate an id for it. While it is not the same thing as vnode boundary, I fail to see how we could have data at  ID ranges outside it for a given vnode. So I still think it should be valid ranges to query.

As for scan vs IN, the latter would require an iterator to know all the ID:s contained in a vnode, and we need to encode this in (hopefully shorter than) 4096 chars. Otherwise we need statefulness, or secondary queries. Whereas a token range is easily encoded in 128bits.

Reply:
I meant a situation like the following: say vnode is `(100, 200]`, and shards have the following ranges:
`(90, 130]` - shard 1
`(130, 170]` - shard 0
`(170, 210]` - shard 1
(I'm not sure which boundaries are closed and which are opened, but let's assume it's the same as for vnodes)

Then, if I understand correctly:
- there will be two stream IDs in this vnode with tokens `131` (for shard 0) and `171` (for shard 1)
- your code will perform a range query `(170, 212)` or something like that (because the first stream ID in the following vnode will have token 211), so we perform a cross-vnode scan.

another approach would be to query `(130, 172)` (or `[131, 171]` or whatever) and simply give up trying to cover the entire token ring with our ranges, because we don't really need to if we just want to cover all stream IDs. Then there won't be any cross-vnode scans.

Reply:
Right, that was what I was a bit worried about - we will potentially query the "wrong" vnode and get zilch. . But as you say, we can eliminate the cross-vnode scan by simply changing the shard range to  be lowest-id-start-inclusive to highest-id-start-inclusive. This will include all ID:s generated for the vnode, and is actually even easier to build.